src/core/account/Account.ts

```diff
@@ -0,0 +1,245 @@
+import type { AccountAuthenticator } from "../../transactions/authenticator/account";
+import { HexInput, SigningScheme, SigningSchemeInput } from "../../types";
+import type { AccountAddress, AccountAddressInput } from "../accountAddress";
+import { AuthenticationKey } from "../authenticationKey";
+import { AccountPublicKey, Ed25519PrivateKey, PrivateKey, Signature, VerifySignatureArgs } from "../crypto";
+import { Ed25519Account } from "./Ed25519Account";
+import { SingleKeyAccount } from "./SingleKeyAccount";
+
+/**
+ * Arguments for creating an `Ed25519Account` from an `Ed25519PrivateKey`.
+ * This is the default input type when passing an `Ed25519PrivateKey`.
+ * In order to use the SingleKey authentication scheme, `legacy` needs to be explicitly set to false.
+ */
+export interface CreateEd25519AccountFromPrivateKeyArgs {
+  privateKey: Ed25519PrivateKey;
+  address?: AccountAddressInput;
+  legacy?: true;
+}
+
+/**
+ * Arguments for creating an `SingleKeyAccount` from an `Ed25519PrivateKey`.
+ * The `legacy` argument needs to be explicitly set to false in order to
+ * use the `SingleKey` authentication scheme.
+ */
+export interface CreateEd25519SingleKeyAccountFromPrivateKeyArgs {
+  privateKey: Ed25519PrivateKey;
+  address?: AccountAddressInput;
+  legacy: false;
+}
+
+/**
+ * Arguments for creating an `SingleKeyAccount` from any supported private key
+ * that is not an `Ed25519PrivateKey`.
+ * The `legacy` argument defaults to false and cannot be explicitly set to true.
+ */
+export interface CreateSingleKeyAccountFromPrivateKeyArgs {
+  privateKey: Exclude<PrivateKey, Ed25519PrivateKey>;
+  address?: AccountAddressInput;
+  legacy?: false;
+}
+
+/**
+ * Arguments for creating an opaque `Account` from any supported private key.
+ * This is used when the private key type is not known at compilation time.
+ */
+export interface CreateAccountFromPrivateKeyArgs {
+  privateKey: PrivateKey;
+  address?: AccountAddressInput;
+  legacy?: boolean;
+}
+
+/**
+ * Arguments for generating an `Ed25519Account`.
+ * This is the input type used by default.
+ */
+export interface GenerateEd25519AccountArgs {
+  scheme?: SigningSchemeInput.Ed25519;
+  legacy?: true;
+}
+
+/**
+ * Arguments for generating an `SingleKeyAccount` with ah underlying `Ed25519PrivateKey`.
+ * The `legacy` argument needs to be explicitly set to false,
+ * otherwise an `Ed25519Account` will be returned instead.
+ */
+export interface GenerateEd25519SingleKeyAccountArgs {
+  scheme?: SigningSchemeInput.Ed25519;
+  legacy: false;
+}
+
+/**
+ * Arguments for generating an `SingleKeyAccount` with any supported private key
+ * that is not an `Ed25519PrivateKey`.
+ * The `legacy` argument defaults to false and cannot be explicitly set to true.
+ */
+export interface GenerateSingleKeyAccountArgs {
+  scheme: Exclude<SigningSchemeInput, SigningSchemeInput.Ed25519>;
+  legacy?: false;
```

Comment:
The ? is so that both of the following are accepted:
```typescript
Account.generate({ SigninSchemeInput.Secp256k1 }): SingleKeyAccount;
Account.generate({ SigninSchemeInput.Secp256k1, legacy: false }): SingleKeyAccount;
```
but this isn't
```typescript
Account.generate({ SigninSchemeInput.Secp256k1, legacy: true }): never;
```

Reply:
i see i see.